server/conn.go

```diff
@@ -1433,7 +1433,19 @@ func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 
 		capabilities := cc.ctx.GetSessionVars().ClientCapability
 		if capabilities&mysql.ClientMultiStatements < 1 {
```

Comment:
Is it intended that, if user set multi-statements on,  even MultiStatementMode is Warn, the server will not report any error? Or if user set multi-statements on,  even MultiStatementMode is Off, the server will continue without errors?

If it is, is there any way to test it?

Reply:
> Is it intended that, if user set multi-statements on, even MultiStatementMode is Warn, the server will not report any error? Or if user set multi-statements on, even MultiStatementMode is Off, the server will continue without errors?

Correct: If multiStatements are ON or WARN, there is no error. The error is caused by OFF, because multiStatement support is "turned off". This is assuming that the client has not set the multi-statement flag correctly, in which case it is permitted and none of this applies.

> That is, is it intended to put your MultiStatementMode code in the branch of multi-statements disabled
> by users/clients?

Correct. If they enable multi-statements in their client, I have no issue with it. This flag is only for preserving the upgrade story because TiDB previously ignored the client multi-statement flag.

> If it is, is there any way to test cases where multi-statements are enabled by clients?

Yes, there are preexisting tests. See `runTestMultiStatements` in `server/server_test.go` which started setting this flag correctly from the PR https://github.com/pingcap/tidb/pull/19459/files

Reply:
Thanks for you detailed explanation :)

Reply:
it that a conclusion above?
client `multi-statements = on` will pass through the check, otherwise, TiDB will take internal MultiStatementMode into consideration?

Reply:
Correct. Ideally clients would always set this, and we don't have the internal MultiStatementMode check. But per the description, this is a workaround because TiDB only started enforcing the multi-statement check in 4.0.9 and it's been causing users problems.